Mathlib/Algebra/Module/ZMod.lean

```diff
@@ -40,6 +40,50 @@ abbrev AddCommGroup.zmodModule {G : Type*} [AddCommGroup G] (h : ∀ (x : G), n
   | 0 => AddCommGroup.toIntModule G
   | _ + 1 => AddCommMonoid.zmodModule h
 
+theorem AddCommGroup.zmodModule.coe_smul {G : Type*} [AddCommGroup G] (h : ∀ (g : G), n • g = 0)
+    (k : ℤ) (g : G) : let _ := AddCommGroup.zmodModule h; (k : ZMod n) • g = k • g :=
+  match n with
+  | 0 => rfl
+  | m + 1 => by
+    have H := (k.mod_modEq (m + 1 : ℕ)).of_dvd
+      (Int.natCast_dvd_natCast.mpr (addOrderOf_dvd_iff_nsmul_eq_zero.mpr (h g)))
+    rwa [← zsmul_eq_zsmul_iff_modEq, ← ZMod.val_intCast, Nat.cast_smul_eq_nsmul] at H
+
+/-- The `ZMod n`-module structure on Abelian groups whose elements have order dividing `n`.
+See note [reducible non-instances]. -/
+abbrev CommGroup.zmodModule {G : Type*} [CommGroup G] (h : ∀ (g : G), g ^ n = 1) :
+    MulDistribMulAction (ZMod n) G :=
+  let _ := (AddCommGroup.zmodModule (G := Additive G) h).toDistribMulAction
+  { smul m a := m • Additive.ofMul a
```

Comment:
I feel like this maybe deserves a type synonym instead...
What do you think?

Reply:
Could you spell out your suggestion a bit more? Are you suggesting a type synonym that bundles `G` with a proof that `g ^ n = 1` and has a natural action of `ZMod n`?

Reply:
(And would you want the same for the existing `AddCommMonoid.zmodModule`?)

Reply:
Yeah, all the subsequent decls need this `let` inside the statement.
And the flipping back and forth between additive and multiplicative seems like it might best be hidden behind a bit more abstraction boundary.

Reply:
I've made an attempt at this refactor, but let me know if anything could be improved.